pkg/transient-disk/transient-disk.go

```diff
@@ -0,0 +1,154 @@
+	var containers []kubev1.Container
+
+	initialDelaySeconds := 2
+	timeoutSeconds := 5
+	periodSeconds := 5
+	successThreshold := 2
+	failureThreshold := 5
+
+	// Make VM Image Wrapper Containers
+	for _, volume := range vm.Spec.Volumes {
+		if volume.Transient != nil {
+			diskContainerName := fmt.Sprintf("transientvolume-%s", volume.Name)
+			diskContainerImage := transientContainerImage
+			filePath, _, _ := GetFilePath(vm, volume.Name)
+
+			baseImageFile, _, err := getFilePathByVolume(vm, volume)
+			if err != nil {
+				return containers, err
+			}
+
+			containers = append(containers, kubev1.Container{
+				Name:            diskContainerName,
+				Image:           diskContainerImage,
+				ImagePullPolicy: kubev1.PullIfNotPresent,
+				Command:         []string{"/entry-point.sh"},
+				Env: []kubev1.EnvVar{
+					{
+						Name:  "BASE_IMAGE",
+						Value: baseImageFile,
+					},
+					{
+						Name:  "COW_IMAGE",
+						Value: filePath,
+					},
+				},
+				VolumeMounts: generateVolumeMountsForVolume(volume, vm),
+				// The readiness probes ensure the volume coversion and copy finished
+				// before the container is marked as "Ready: True"
+				ReadinessProbe: &kubev1.Probe{
+					Handler: kubev1.Handler{
+						Exec: &kubev1.ExecAction{
+							Command: []string{
+								"cat",
+								"/tmp/healthy",
+							},
+						},
+					},
+					InitialDelaySeconds: int32(initialDelaySeconds),
+					PeriodSeconds:       int32(periodSeconds),
+					TimeoutSeconds:      int32(timeoutSeconds),
+					SuccessThreshold:    int32(successThreshold),
+					FailureThreshold:    int32(failureThreshold),
+				},
+			})
+		}
+	}
+
+	return containers, nil
+}
+
+func GenerateVolumes(vm *v1.VirtualMachine) ([]kubev1.Volume, error) {
+	var volumes []kubev1.Volume
+
+	for _, volume := range vm.Spec.Volumes {
+		if volume.Transient != nil {
+			volumes = append(volumes, kubev1.Volume{
+				Name: volume.Name,
+				VolumeSource: kubev1.VolumeSource{
+					EmptyDir: &kubev1.EmptyDirVolumeSource{},
+				},
+			})
+			volumes = append(volumes, kubev1.Volume{
+				Name: createBackingVolumeName(volume.Name),
+				VolumeSource: kubev1.VolumeSource{
+					PersistentVolumeClaim: volume.Transient.PersistentVolumeClaim,
+				},
+			})
+		}
+	}
+
+	return volumes, nil
+}
+
+func GenerateVolumeMounts(vm *v1.VirtualMachine) []kubev1.VolumeMount {
+	var volumeMounts []kubev1.VolumeMount
+
+	for _, volume := range vm.Spec.Volumes {
+		if volume.Transient != nil {
+			if volume.Transient != nil {
+				volumeMounts = append(volumeMounts, generateVolumeMountsForVolume(volume, vm)...)
+			}
+		}
+	}
+
+	return volumeMounts
+}
+
+func generateVolumeMountsForVolume(volume v1.Volume, vm *v1.VirtualMachine) []kubev1.VolumeMount {
+	var volumeMounts []kubev1.VolumeMount
+
+	volumeMounts = append(volumeMounts, kubev1.VolumeMount{
+		Name:      createBackingVolumeName(volume.Name),
+		MountPath: filepath.Join("/var/run/kubevirt-private", "vm-disks", volume.Name),
```

Comment:
I'd like to see all this logic involved with building PVC file paths converged into a single location. We're building this path in template.go, converter.go, and here too now.

Reply:
Oh yeah totally, I was thinking of pkg/pvc -- symmetric to pkg/transientdisk and pkg/registrydisk, where each "storage type" provides interface that the template uses. This case is the worst offender.